### PR TITLE
Sabre

### PR DESCRIPTION
**Pull request by Simone-Bordoni**
Add Sabre implementation (sabre_from_scratch.py) to qibolab as routing algorithm.

Checklist:
- [x] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

### CODE REVIEW

**Review by stavros11 — COMMENTED** [review]
Thanks for implementing this @Simone-Bordoni. Have you tried to use this on a circuit that contains measurements? For example I tried the following
```py
from qibo import gates, Circuit
from qibolab import create_platform
from qibolab.transpilers.pipeline import Passes
from qibolab.transpilers.router import Sabre
from qibolab.backends import QibolabBackend

platform = create_platform("dummy")

transpiler = Passes(connectivity=platform.topology)
transpiler.passes[2] = Sabre(platform.topology)

c = Circuit(2)
c.add(gates.H(0))
c.add(gates.CZ(0, 1))
c.add(gates.M(0, 1))

tc, final_layout = transpiler(c)

print(c.draw())
print()
print(tc.draw())
```
but it failed with
```
AttributeError: 'M' object has no attribute '_control_qubits'. Did you mean: 'control_qubits'?
```
I thought that this is an issue with qibo, but when I tried to just deepcopy the qibo circuit it worked. Maybe the `CircuitMap` is doing a weird manipulation that is causing a bug in qibo, could you please have a look?

**Comment by Simone-Bordoni** [comment]
> Thanks for implementing this @Simone-Bordoni. Have you tried to use this on a circuit that contains measurements? For example I tried the following
> 
> ```python
> from qibo import gates, Circuit
> from qibolab import create_platform
> from qibolab.transpilers.pipeline import Passes
> from qibolab.transpilers.router import Sabre
> from qibolab.backends import QibolabBackend
> 
> platform = create_platform("dummy")
> 
> transpiler = Passes(connectivity=platform.topology)
> transpiler.passes[2] = Sabre(platform.topology)
> 
> c = Circuit(2)
> c.add(gates.H(0))
> c.add(gates.CZ(0, 1))
> c.add(gates.M(0, 1))
> 
> tc, final_layout = transpiler(c)
> 
> print(c.draw())
> print()
> print(tc.draw())
> ```
> 
> but it failed with
> 
> ```
> AttributeError: 'M' object has no attribute '_control_qubits'. Did you mean: 'control_qubits'?
> ```
> 
> I thought that this is an issue with qibo, but when I tried to just deepcopy the qibo circuit it worked. Maybe the `CircuitMap` is doing a weird manipulation that is causing a bug in qibo, could you please have a look?

You are right I did not take into consideration measurements...
Actually I should also include them in the block decomposition but i am not sure about how to handle them.
If you want the following days i will try to find a solution, otherwise i will fix this issue after moving it to qibo.

File: src/qibolab/transpilers/blocks.py [hunk 1]
```diff
@@ -92,6 +102,56 @@ def kak_decompose(self):  # pragma: no cover
...
+    def __init__(self, circuit: Circuit, index_names: bool = False):
+        self.block_list = block_decomposition(circuit)
+        self._index_names = index_names
+        if index_names:
+            for index, block in enumerate(self.block_list):
+                block.rename(index)
+        self.qubits = circuit.nqubits
+
+    def __call__(self):
+        return self.block_list
```
[the remark this answers is not in the document]
Comment (Simone-Bordoni) [reply src/qibolab/transpilers/blocks.py:122]: Ok, if you want I can change it to return the i-th element. But i mainly use this function in sabre where i need to return the full block decomposition as a list.

File: src/qibolab/transpilers/placer.py [hunk 2]
```diff
@@ -150,10 +150,8 @@ def __call__(self, circuit: Circuit):
                 self.connectivity.number_of_edges() == circuit_subgraph.number_of_edges()
                 or i == len(gates_qubits_pairs) - 1
             ):
-                keys = list(result.mapping.keys())
-                keys.sort()
-                return {i: result.mapping[i] for i in keys}
-        return dict(sorted(result.mapping.items()))
+                return {"q" + str(i): result.mapping[i] for i in range(len(result.mapping))}
+        return {"q" + str(i): result.mapping[i] for i in range(len(result.mapping))}
```
[the remark this answers is not in the document]
Comment (Simone-Bordoni) [reply src/qibolab/transpilers/placer.py:154]: But in the first return i break the while cicle. I will replace it with a "break"

**Review by BrunoLiegiBastonLiegi — COMMENTED** [review]
Thanks @Simone-Bordoni , it's looking good. I would add the test for checking that the final state of the transpiled circuit is the same of the original one.

**Comment by Simone-Bordoni** [comment]
I have added a test to check circuit equivalence from evolving a quantum state. This test may be useful for every transpiler.
As StarConnectivity was realized before, it has not been possible to make it work with the new test, in the future if we will decide to keep it I will rewrite it to make it work.
If you agree with the modifications we can merge and move all the transpiler folder to qibo.

File: tests/test_transpilers_router.py [hunk 3]
```diff
@@ -171,48 +157,52 @@ def test_random_circuits_5q(gates, qubits):
     transpiler = ShortestPaths(connectivity=star_connectivity(), verbose=True)
     circuit = generate_random_circuit(nqubits=qubits, ngates=gates)
     transpiled_circuit, final_qubit_map = transpiler(circuit, initial_layout)
-    assert transpiler.added_swaps >= 0
+    assert transpiler.added_swaps() >= 0
```
Comment (BrunoLiegiBastonLiegi) [inline tests/test_transpilers_router.py:144]: Probably better to make this a `@property`

File: tests/test_transpilers_router.py [hunk 4]
```diff
@@ -171,48 +157,52 @@ def test_random_circuits_5q(gates, qubits):
...
     initial_layout = placer(star_circuit())
     transpiler = ShortestPaths(connectivity=star_connectivity())
     transpiled_circuit, final_qubit_map = transpiler(star_circuit(), initial_layout)
-    assert transpiler.added_swaps == 0
+    assert transpiler.added_swaps() == 0
     assert_connectivity(star_connectivity(), transpiled_circuit)
     assert_placement(transpiled_circuit, final_qubit_map)
-    assert final_qubit_map["q2"] == 0
+    print(initial_layout)
+    print(final_qubit_map)
```
Comment (BrunoLiegiBastonLiegi) [inline tests/test_transpilers_router.py:167]:
```suggestion
```
I would remove these print

**Comment by BrunoLiegiBastonLiegi** [comment]
Thanks @Simone-Bordoni. I proposed just some minor corrections.
I noticed, that all the tests use the `Trivial` placer and the `star_connectivity`, might be good to test at least a couple of different ones. Maybe set them as pytest parameters.

**Review by BrunoLiegiBastonLiegi — APPROVED** [review]
Thanks @Simone-Bordoni , everything seems to be fine to me now.

**Review by stavros11 — APPROVED** [review]

**Pull request merged**